From 06af5ac88a95e9ccc2fcd5faacff1f103c356afd Mon Sep 17 00:00:00 2001
From: chickendrop89 <chickendrop89@gmail.com>
Date: Fri, 9 May 2025 23:29:59 +0200
Subject: [PATCH] kernel: printk: add a way of suppressing GKI module debugging

* Some vendor GKI modules were built with debugging compiled in, this
spams the console with info we won't need and covers up important logs.

* More information in Kconfig

[Idea from this commit by pzqqt] https://github.com/Pzqqt/android_kernel_xiaomi_marble/commit/aab4791f03e0824fd68ba0f257672959e1134d7c

Signed-off-by: chickendrop89 <chickendrop89@gmail.com>
---
 init/Kconfig           | 15 +++++++++++++
 kernel/printk/printk.c | 50 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 65 insertions(+)

diff --git a/init/Kconfig b/init/Kconfig
index 3aef1d9111af..655745c52e3b 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2414,6 +2414,21 @@ config UNUSED_KSYMS_WHITELIST
 	  one per line. The path can be absolute, or relative to the kernel
 	  source tree.
 
+config SUPPRESS_VENDOR_MODULE_DEBUGGING
+	bool "Suppress vendor module debugging"
+	help
+	  Some vendor GKI modules were built with debugging compiled in, this
+	  spams the console with info we won't need and covers up important
+	  logs.
+
+	  This hack is specific to xiaomi devices. The modules affected can 
+	  be found in the filtered_modules[] array in kernel/printk.c
+
+	  Do not select this if you are debugging, or planning to build the 
+	  modules into the kernel image.
+
+	  If unsure, say N.
+
 endif # MODULES
 
 config MODULES_TREE_LOOKUP
diff --git a/kernel/printk/printk.c b/kernel/printk/printk.c
index 58929f2d986c..10fa60793643 100644
--- a/kernel/printk/printk.c
+++ b/kernel/printk/printk.c
@@ -2142,6 +2142,44 @@ static u16 printk_sprint(char *text, u16 size, int facility,
 	return text_len;
 }
 
+__maybe_unused static bool should_filter_vendor_message(const char *text) 
+{
+    const char *const *module;
+
+	/* 
+	 * The fmt string in question needs to be truncated to 5
+	 * characters due to the prefix_buf buffer being 8 bytes 
+	 */
+	static const char *const filtered_modules[] = {
+        #if !IS_ENABLED(CONFIG_POWER_SUPPLY_DEBUG)
+            "[bq25", /* [bq2589x] */
+            "[sm56", /* [sm5602]  */
+            "nopmi", /* nopmi_chg */
+            "[sc85", /* [sc8551-STANDALONE]*/
+        #endif
+            "st21n", /* st21nfc  */
+            "[Awin", /* [Awinic] */
+            NULL
+    };
+
+    for (module = filtered_modules; likely(*module); module++) {
+        /* 
+         * The PSU logs are the largest, and most repeated portion 
+         * of what is printed in the console most of the time. 
+         * Adjust branch perdiction accordingly  
+         */
+        #if !IS_ENABLED(CONFIG_POWER_SUPPLY_DEBUG)
+            if (likely(strstr(text, *module)))
+                return true;
+        #else
+            if (unlikely(strstr(text, *module)))
+                return true;
+        #endif
+    }
+    
+    return false;
+}
+
 __printf(4, 0)
 int vprintk_store(int facility, int level,
 		  const struct dev_printk_info *dev_info,
@@ -2182,6 +2220,18 @@ int vprintk_store(int facility, int level,
 	reserve_size = vsnprintf(&prefix_buf[0], sizeof(prefix_buf), fmt, args2) + 1;
 	va_end(args2);
 
+	/* 
+	 * Some vendor GKI modules were built with debugging compiled in, 
+	 * this spams the console with info we won't need and covers up 
+	 * important logs
+	 */
+	#if IS_ENABLED(CONFIG_SUPPRESS_VENDOR_MODULE_DEBUGGING)
+        if (should_filter_vendor_message(prefix_buf)) {
+           ret = 0;
+           goto out;
+        }
+	#endif
+
 	if (reserve_size > LOG_LINE_MAX)
 		reserve_size = LOG_LINE_MAX;
 
